File: Python/Trees/Binary_Tree_View.py
from collections import deque

class Node:
    def __init__(self, k):
        self.key = k
        self.left = None
        self.right = None

def insert(root):
    print("Enter value to be entered: ", end="")
    val = int(input())
    root = Node(val)

    if val == -1:
        return None

    print("Enter value to the left of", val, ": ", end="")
    root.left = insert(root.left)
    print("Enter value to the right of", val, ": ", end="")
    root.right = insert(root.right)

    return root

def display(V):
    for i in V:
        print(i, end=" ")

def left_view(root, level, res):
    if not root:
        return
    if len(res) == level:
        res.append(root.key)

    left_view(root.left, level + 1, res)
    left_view(root.right, level + 1, res)

def right_view(root, level, res):
    if not root:
        return
    if len(res) == level:
        res.append(root.key)

    right_view(root.right, level + 1, res)
    right_view(root.left, level + 1, res)

def top_view(root):
    res = []
    if not root:
        return res

    mp = {}
    q = deque([(root, 0)])
    while q:
        f = q.popleft()
        node, line = f[0], f[1]

        if line not in mp:
            mp[line] = node.key

        if node.left:
            q.append((node.left, line - 1))
        if node.right:
            q.append((node.right, line + 1))

    for i in sorted(mp.keys()):
        res.append(mp[i])

    return res

def bottom_view(root):
    res = []
    if not root:
        return res

    mp = {}
    q = deque([(root, 0)])
    while q:
        f = q.popleft()
        node, line = f[0], f[1]

        mp[line] = node.key

        if node.left:
            q.append((node.left, line - 1))
        if node.right:
            q.append((node.right, line + 1))

    for i in sorted(mp.keys()):
        res.append(mp[i])

    return res

if __name__ == "__main__":
    root = None
    print("Enter binary tree nodes, and for no node, enter -1\n")
    root = insert(root)
    L, R, T, B = [], [], [], []
    usrchoice = 0

    while True:
        print("\n\n1. Left View")
        print("2. Right View")
        print("3. Top View")
        print("4. Bottom View")
        print("5. Exit")
        print("Enter the choice of operation to be implemented: ", end="")
        usrchoice = int(input())

        if usrchoice == 1:
            left_view(root, 0, L)
            display(L)
        elif usrchoice == 2:
            right_view(root, 0, R)
            display(R)
        elif usrchoice == 3:
            T = top_view(root)
            display(T)
        elif usrchoice == 4:
            B = bottom_view(root)
            display(B)
        elif usrchoice == 5:
            break
        else:
            print("Enter a valid choice")

File: Java/Trees/Binary_Tree_View.java
import java.util.*;
import java.util.Scanner;

class Binary_Tree_View {
    static Scanner scanner = new Scanner(System.in);

    static class Node {
        int key;
        Node left, right;
        int level;

        // Constructor
        Node(int k) {
            key = k;
            left = right = null;
        }
    }

    static Node insert(Node root) {
        System.out.print("Enter value to be entered: ");
        int val = scanner.nextInt();
        root = new Node(val);

        if (val == -1)
            return null;

        System.out.print("Enter value to left of " + val + ": ");
        root.left = insert(root.left);
        System.out.print("Enter value to right of " + val + ": ");
        root.right = insert(root.right);

        return root;
    }

    static void display(List<Integer> V) {
        for (int i : V) {
            System.out.print(i + " ");
        }
    }

    static class Pair<T, U> {
        public T first;
        public U second;

        public Pair(T first, U second) {
            this.first = first;
            this.second = second;
        }
    }

    static List<Integer> leftView(Node root) {
        List<Integer> res = new ArrayList<>();
        if (root == null)
            return res;

        // initialising variables
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        root.level = 0;
        int level;

        // Tree Map keeps entries of Map sorted by key
        TreeMap<Integer, Integer> mp = new TreeMap<>();
        
        while (!q.isEmpty()) 
        {
            Node temp = q.peek();
            level = temp.level;

            if(!mp.containsKey(level))
                mp.put(level, temp.key);
            
            q.poll();

            if (temp.left != null) 
            {
                temp.left.level = level + 1;
                q.add(temp.left);
            }
            if (temp.right != null) 
            {
                temp.right.level = level + 1;
                q.add(temp.right);
            }    
        }
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            res.add(entry.getValue());
        }
        return res;
    }

    static List<Integer> rightView(Node root) {
        List<Integer> res = new ArrayList<>();
        if (root == null)
            return res;

        Queue<Node> q = new LinkedList<>();
        q.add(root);
        root.level = 0;
        int level;
        TreeMap<Integer, Integer> mp = new TreeMap<>();

        while (!q.isEmpty()) 
        {
            Node temp = q.peek();
            level = temp.level;

            mp.put(level, temp.key);
            q.poll();

            if (temp.left != null) 
            {
                temp.left.level = level + 1;
                q.add(temp.left);
            }
            if (temp.right != null) 
            {
                temp.right.level = level + 1;
                q.add(temp.right);
            }   
        }
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            res.add(entry.getValue());
        }
        return res;
    }

    static List<Integer> topView(Node root) {
        List<Integer> res = new ArrayList<>();
        if (root == null)
            return res;

        Queue<Node> q = new LinkedList<>();
        q.add(root);
        root.level = 0;
        TreeMap<Integer, Integer> mp = new TreeMap<>();
        int level;

        while (!q.isEmpty()) 
        {
            Node temp = q.peek();
            level = temp.level;

            if (!mp.containsKey(level))
                mp.put(level, temp.key);
            q.poll();

            if (temp.left != null) 
            {
                temp.left.level = level - 1;
                q.add(temp.left);
            }
            if (temp.right != null) 
            {
                temp.right.level = level + 1;
                q.add(temp.right);
            }
        }
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            res.add(entry.getValue());
        }
        return res;
    }

    static List<Integer> bottomView(Node root) {
        List<Integer> res = new ArrayList<>();
        
        if (root == null)
            return res;

        Queue<Node> q = new LinkedList<>();
        q.add(root);
        root.level = 0;
        int level;
        TreeMap<Integer, Integer> mp = new TreeMap<>();
       
        while (!q.isEmpty()) 
        {
            Node temp = q.peek();
            level = temp.level;

            mp.put(level, temp.key);
            q.poll();

            if (temp.left != null) 
            {
                temp.left.level = level - 1;
                q.add(temp.left);
            }
            if (temp.right != null) 
            {
                temp.right.level = level + 1;
                q.add(temp.right);
            }
        }
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            res.add(entry.getValue());
        }

        return res;
    }

    public static void main(String[] args) {
        Node root = null;
        System.out.println("Enter binary tree nodes, and for no node enter -1\n");
        root = insert(root);
        List<Integer> L = new ArrayList<>();
        List<Integer> R = new ArrayList<>();
        List<Integer> T = new ArrayList<>();
        List<Integer> B = new ArrayList<>();
        int usrchoice = 0;

        while (true) {
            System.out.println("\n\n1.Left View\n2.Right View\n3.Top View\n4.Bottom View\n5.Exit");
            System.out.print("Enter the choice of operation to be implemented: ");
            usrchoice = scanner.nextInt();

            switch (usrchoice) {
                case 1:
                    L=leftView(root);
                    display(L);
                    break;

                case 2:
                    R=rightView(root);
                    display(R);
                    break;

                case 3:
                    T = topView(root);
                    display(T);
                    break;

                case 4:
                    B = bottomView(root);
                    display(B);
                    break;

                case 5:
                    System.exit(0);

                default:
                    System.out.print("Enter a valid choice: ");
                    break;
            }
        }
    }
}